hack/generator/pkg/astmodel/file_definition_test.go

```diff
@@ -11,24 +11,144 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+/*
+ * NewFileDefinition tests
+ */
+
 func Test_NewFileDefinition_GivenValues_InitializesFields(t *testing.T) {
 	g := NewGomegaWithT(t)
 
-	person := NewTestStruct("Person", "fullName", "knownAs", "familyName")
+	person := NewTestStruct(
+		"Person",
+		NewStringPropertyDefinition("fullName"),
+		NewStringPropertyDefinition("knownAs"),
+		NewStringPropertyDefinition("familyName"),
+	)
 	file := NewFileDefinition(&person.Name().PackageReference, &person)
 
 	g.Expect(*file.packageReference).To(Equal(person.Name().PackageReference))
 	g.Expect(file.definitions).To(HaveLen(1))
 }
 
-func NewTestStruct(name string, properties ...string) StructDefinition {
-	var ps []*PropertyDefinition
-	for _, n := range properties {
-		ps = append(ps, NewPropertyDefinition(PropertyName(n), n, StringType))
-	}
+/*
+ * calcRanks() tests
+ */
+
+func Test_CalcRanks_GivenMultipleRoots_AssignsRankZeroToAll(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	root1 := NewTestStruct("r1")
+	root2 := NewTestStruct("b")
+	root3 := NewTestStruct("c")
+	root4 := NewTestStruct("d")
+
+	ranks := calcRanks([]TypeDefiner{&root1, &root2, &root3, &root4})
+
+	g.Expect(ranks[*root1.Name()]).To(Equal(0))
+	g.Expect(ranks[*root2.Name()]).To(Equal(0))
+	g.Expect(ranks[*root3.Name()]).To(Equal(0))
+	g.Expect(ranks[*root4.Name()]).To(Equal(0))
+}
+
+func Test_CalcRanks_GivenLinearDependencies_AssignsRanksInOrder(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	rank3 := NewTestStruct("d")
+	referenceToRank3 := NewPropertyDefinition("f3", "f3", rank3.Name())
+
+	rank2 := NewTestStruct("c", referenceToRank3)
+	referenceToRank2 := NewPropertyDefinition("f2", "f2", rank2.Name())
+
+	rank1 := NewTestStruct("b", referenceToRank2)
+	referenceToRank1 := NewPropertyDefinition("f1", "f1", rank1.Name())
+
+	rank0 := NewTestStruct("a", referenceToRank1)
+
+	ranks := calcRanks([]TypeDefiner{&rank0, &rank1, &rank2, &rank3})
+
+	g.Expect(ranks[*rank0.Name()]).To(Equal(0))
+	g.Expect(ranks[*rank1.Name()]).To(Equal(1))
+	g.Expect(ranks[*rank2.Name()]).To(Equal(2))
+	g.Expect(ranks[*rank3.Name()]).To(Equal(3))
+}
+
+func Test_CalcRanks_GivenDiamondDependencies_AssignRanksInOrder(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	bottom := NewTestStruct("bottom")
+	referenceToBottom := NewPropertyDefinition("b", "b", bottom.Name())
+
+	left := NewTestStruct("l", referenceToBottom)
+	referenceToLeft := NewPropertyDefinition("l", "l", left.Name())
+
+	right := NewTestStruct("r", referenceToBottom)
+	referenceToRight := NewPropertyDefinition("r", "r", right.Name())
+
+	top := NewTestStruct("a", referenceToLeft, referenceToRight)
+
+	ranks := calcRanks([]TypeDefiner{&top, &left, &right, &bottom})
+
+	g.Expect(ranks[*top.Name()]).To(Equal(0))
+	g.Expect(ranks[*right.Name()]).To(Equal(1))
+	g.Expect(ranks[*left.Name()]).To(Equal(1))
+	g.Expect(ranks[*bottom.Name()]).To(Equal(2))
+}
+
+func Test_CalcRanks_GivenDiamondWithBar_AssignRanksInOrder(t *testing.T) {
```

Comment:
I'm not sure what `DiamondWithBar` and `DiamondWithReverseBar` mean - might be worth a couple of ascii-art diagram comments?

Reply:
Done.